File: sdk/scala-sdk-protobuf/src/main/scala/kalix/scalasdk/action/MessageEnvelope.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.scalasdk.action

import kalix.scalasdk.Metadata
import kalix.scalasdk.impl.action.MessageEnvelopeImpl

object MessageEnvelope {

  /**
   * Create a message.
   *
   * @param payload
   *   The payload of the message.
   * @return
   *   The message.
   */
  def apply[T](payload: T): MessageEnvelope[T] = new MessageEnvelopeImpl[T](payload, Metadata.empty)

  /**
   * Create a message.
   *
   * @param payload
   *   The payload of the message.
   * @param metadata
   *   The metadata associated with the message.
   * @return
   *   The message.
   */
  def apply[T](payload: T, metadata: Metadata): MessageEnvelope[T] = new MessageEnvelopeImpl[T](payload, metadata)
}
trait MessageEnvelope[T] {

  /**
   * The metadata associated with the message.
   *
   * @return
   *   The metadata.
   */
  def metadata: Metadata

  /**
   * The payload of the message.
   *
   * @return
   *   The payload.
   */
  def payload: T

}

File: sdk/java-sdk-protobuf/src/main/java/kalix/javasdk/action/MessageEnvelope.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk.action;

import kalix.javasdk.Metadata;
import kalix.javasdk.impl.action.MessageEnvelopeImpl;

/** A message envelope. */
public interface MessageEnvelope<T> {
  /**
   * The metadata associated with the message.
   *
   * @return The metadata.
   */
  Metadata metadata();

  /**
   * The payload of the message.
   *
   * @return The payload.
   */
  T payload();

  /**
   * Create a message.
   *
   * @param payload The payload of the message.
   * @return The message.
   */
  static <T> MessageEnvelope<T> of(T payload) {
    return new MessageEnvelopeImpl<>(payload, Metadata.EMPTY);
  }

  /**
   * Create a message.
   *
   * @param payload The payload of the message.
   * @param metadata The metadata associated with the message.
   * @return The message.
   */
  static <T> MessageEnvelope<T> of(T payload, Metadata metadata) {
    return new MessageEnvelopeImpl<>(payload, metadata);
  }
}
